plugin-server/src/utils/db/db.ts

```diff
@@ -529,6 +540,123 @@ export class DB {
+            return PersonPropertyUpdateOperation.Set
+        }
+        return person.properties_last_operation[key]
+    }
+
+    private updatePropertiesLastOperation(person: Person, key: string, value: PersonPropertyUpdateOperation) {
+        // to get around the type check
+        if (person.properties_last_operation === null) {
+            person.properties_last_operation = {}
+        }
+        person.properties_last_operation[key] = value
+    }
+
+    private updatePersonPropertiesLocal(
+        person: Person,
+        properties: Properties,
+        propertiesOnce: Properties,
+        timestamp: DateTime
+    ): boolean {
+        // updates the person & returns true/false if anything was updated
+        let updatedSomething = false
+        Object.entries(propertiesOnce).forEach(([key, value]) => {
+            if (
+                !(key in person.properties) ||
+                (this.getPropertiesLastOperationOrSet(person, key) == PersonPropertyUpdateOperation.SetOnce &&
+                    this.getPropertyLastUpdatedAtDateTimeOrEpoch(person, key) > timestamp)
+            ) {
+                updatedSomething = true
+                person.properties[key] = value
+                this.updatePropertiesLastOperation(person, key, PersonPropertyUpdateOperation.SetOnce)
+                person.properties_last_updated_at[key] = timestamp.toISO()
+            }
+        })
+        // note that if the key appears twice we override it with set value here
+        Object.entries(properties).forEach(([key, value]) => {
+            if (
+                !(key in person.properties) ||
+                this.getPropertiesLastOperationOrSet(person, key) == PersonPropertyUpdateOperation.SetOnce ||
+                this.getPropertyLastUpdatedAtDateTimeOrEpoch(person, key) < timestamp
+            ) {
+                updatedSomething = true
+                person.properties[key] = value
+                this.updatePropertiesLastOperation(person, key, PersonPropertyUpdateOperation.Set)
+                person.properties_last_updated_at[key] = timestamp.toISO()
+            }
+        })
+        return updatedSomething
+    }
+
+    public async updatePersonProperties(
+        teamId: number,
+        distinctId: string,
+        properties: Properties,
+        propertiesOnce: Properties,
+        timestamp: DateTime
+    ): Promise<void> {
+        if (properties.length === 0 && propertiesOnce.length === 0) {
+            return
+        }
+        let person: Person | undefined
+        await this.postgresTransaction(async (client) => {
+            person = await this.fetchPerson(teamId, distinctId, client, true)
+            if (!person) {
+                throw new Error(
+                    `Could not find person with distinct id "${distinctId}" in team "${teamId}" to update props`
+                )
+            }
+
+            if (!this.updatePersonPropertiesLocal(person, properties, propertiesOnce, timestamp)) {
```

Comment:
nit: I personally _really_ don't like this pattern of a function call in a conditional modifying an object in memory. 

How about something like:

```js
const [person, shouldUpdate] = this.updatePersonPropertiesLocal(person, properties, propertiesOnce, timestamp)
if (!shouldUpdate) {
    return
}

await this.updatePersonSetProperties(person, client)
```

Reply:
If we want the person to be immutable we'd need to make a deep copy of it, which could be slow with lots of properties and we run this inside a transaction. I know I could inline the code instead, but then we have 200+ line functions which are hard to read.